### DISCUSSION TITLE
Lesson 4: Unable to install web3 despite having Visual C++, Win10 (SOLVED)

### DISCUSSION THREAD

**Discussion by Verde888** [smartcontractkit/full-blockchain-solidity-course-py#335] [Q&A] [answered]
Hi, 
I'm having a problem installing the web3.py.  After  typing `pip install web3` at some point I get the following error. It says:
`error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/`

It appears when trying to install _cytoolz_. 
Same error appears when I try to install it separately with `pip install cytoolz`

I have installed the Visual C++ Redistributable 14 but it doesn't change anything.

A similar problem has been described there, too, but not solved (even people from this course have posted there):
_https://github.com/ethereum/web3.py/issues/2147_

Thanks in advance, impossible to move forward due to this...

**Comment by cromewar** [comment, marked as answer]
Hello @Verde888 as you mentioned in another post, didn't [this](https://web3py.readthedocs.io/en/stable/troubleshooting.html#why-am-i-getting-visual-c-or-cython-not-installed-error) helped?.

Anyways you would have to also install visual studio and the Python development tools, or there is a more efficient solution which is use WSL (windows subsystem for linux).

**Reply by Verde888** [reply]
Hi, unfortunately, having Visual C++ 14 installed does not help. I have not installed the whole Studio IDE (it's huge - I'm not sure which part should I install, its modular). I'm lost.  

So are you suggesting installing running VSCode on WSL?

**Reply by cromewar** [reply]
Well you might install the base Visual Studio editor and the Python Development module, but as you said it's huge, heavy and totally not worth it (you can do it anyway if you want to use 100% windows) or as I said use WSL which already comes with Python ready for web3, however take in consideration if you never used linux before, it'll be a learning curve until you get along with WSL.

![image](https://user-images.githubusercontent.com/20199180/141013141-fafe1511-27bd-4276-9db4-f2e281137e8b.png)

**Reply by Verde888** [reply]
Thank you @cromewar! I will consider the pros and cons and try something.  
EDIT: I'm having some issues with WLS so I decided to install the Python module of Visual Studio (around 8 GB) and it WORKED. I had no problems with finishing the web3.py installation. Thank you! I will consider moving to WSL in the future.

But is this the only choice for a Windows user? Why no one else encounters this problem with missing Visual C++?

**Reply by cromewar** [reply]
Depend a lot of what libraries you have installed on your windows machine, personally I think is a pain to handle all python stuff using windows (as this things could happen frequently) so that's why I work almost exclusively using WSL, no more headaches for me haha.

**Reply by sanspeur33** [reply]
Stuck at same spot in video, installing Web3.

Running setup.py install for cytoolz ... error
    ERROR: Command errored out with exit status 1:

Searching I found that there may be an issue with python 3.10.0?
Really enjoy course, but I hate being stuck on installs, but what I am finding in this crypto development, that understanding the installs/versions/compatibilities is 99% of the work.  The coding itself is simple.

pip version 21.3.1
python 3.10.0
Win10 Pro. 
Installed Visual C++ 14 (searched but have not found way to verify from command prompt) 

ERROR: Command errored out with exit status 1: 'C:\Users\ASutherland\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\ASutherland\\AppData\\Local\\Temp\\pip-install-zjl4t120\\cytoolz_a2c2c4c0bc304aac993c64a7b132ec98\\setup.py'"'"'; __file__='"'"'C:\\Users\\ASutherland\\AppData\\Local\\Temp\\pip-install-zjl4t120\\cytoolz_a2c2c4c0bc304aac993c64a7b132ec98\\setup.py'"'"';f 
= getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\ASutherland\AppData\Local\Temp\pip-record-fudb93is\install-record.txt' --single-version-externally-managed --user --prefix= --compile --install-headers 'C:\Users\ASutherland\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\Include\cytoolz' Check the logs for full command output.

**Reply by cromewar** [reply]
@sanspeur33  did you install visual studio and the python development tools?

**Reply by sanspeur33** [reply]
Yes.  Went back to visual studio after posting installer after reading this 2x and posting my??, and selected all of the C++ build tools.  rebooted and it worked, well I was able to finish the pip install web3 with some warnings.  So I will proceed.  Thank you for getting back so fast.

**Reply by cromewar** [reply]
I'm glad your problem is solve @sanspeur33, best regards.

**Reply by risktakereth** [reply]
I have the same issue @sanspeur33 , but I don't understand how you fixed it. Could you explain me ?

**Reply by risktakereth** [reply]
Fixed : I used this video ^^
